File: csharp/src/CodexWebMcp/Prompts.cs
using System;
using System.Collections.Concurrent;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace CodexWebMcp;

public static class Prompts
{
    private static string? _resolvedDir;
    private static readonly object _dirLock = new();
    private static readonly ConcurrentDictionary<string, string> _templateCache = new();

    public static string ResolveDir()
    {
        if (_resolvedDir is not null) return _resolvedDir;
        lock (_dirLock)
        {
            if (_resolvedDir is not null) return _resolvedDir;

            var attempted = new List<string>();

            // 1. Env var override
            var env = Environment.GetEnvironmentVariable("CODEX_WEB_PROMPTS");
            if (!string.IsNullOrWhiteSpace(env))
            {
                attempted.Add($"env:CODEX_WEB_PROMPTS={env}");
                if (Directory.Exists(env))
                {
                    _resolvedDir = env;
                    return _resolvedDir;
                }
            }

            // 2. Walk up from AppContext.BaseDirectory
            if (TryWalkUp(AppContext.BaseDirectory, attempted, out var dir1))
            {
                _resolvedDir = dir1;
                return _resolvedDir!;
            }

            // 3. Walk up from current working directory
            if (TryWalkUp(Directory.GetCurrentDirectory(), attempted, out var dir2))
            {
                _resolvedDir = dir2;
                return _resolvedDir!;
            }

            var sb = new StringBuilder();
            sb.AppendLine("Could not locate prompts/ directory. Attempted paths:");
            foreach (var a in attempted) sb.AppendLine($"  - {a}");
            throw new InvalidOperationException(sb.ToString());
        }
    }

    private static bool TryWalkUp(string start, List<string> attempted, out string? found)
    {
        found = null;
        if (string.IsNullOrWhiteSpace(start)) return false;

        var cur = Path.GetFullPath(start);
        for (int i = 0; i < 10; i++)
        {
            var candidate = Path.Combine(cur, "prompts");
            attempted.Add(candidate);
            if (Directory.Exists(candidate))
            {
                found = candidate;
                return true;
            }
            var parent = Path.GetDirectoryName(cur);
            if (string.IsNullOrEmpty(parent) || parent == cur) break;
            cur = parent;
        }
        return false;
    }

    public static string LoadTemplate(string toolName)
    {
        return _templateCache.GetOrAdd(toolName, name =>
        {
            var dir = ResolveDir();
            var path = Path.Combine(dir, $"{name}.md");
            if (!File.Exists(path))
                throw new FileNotFoundException(
                    $"Prompt template not found: {path}", path);
            return File.ReadAllText(path);
        });
    }

    public static string Render(string toolName, IReadOnlyDictionary<string, string> vars)
    {
        var template = LoadTemplate(toolName);
        var result = template;
        foreach (var kv in vars)
        {
            result = result.Replace("{" + kv.Key + "}", kv.Value);
        }
        return result;
    }
}

File: java/src/main/java/codexwebmcp/Prompts.java
package codexwebmcp;

import java.io.IOException;
import java.net.URISyntaxException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public final class Prompts {
    private static final Object DIR_LOCK = new Object();
    private static volatile Path dirValue;
    private static final ConcurrentHashMap<String, String> TPL_CACHE = new ConcurrentHashMap<>();

    private Prompts() {}

    private static boolean isPromptsDir(Path p) {
        return Files.isDirectory(p);
    }

    private static Path walkUp(Path start, List<String> attempted) {
        Path cur = start;
        for (int i = 0; i < 10; i++) {
            Path cand = cur.resolve("prompts");
            attempted.add(cand.toString());
            if (isPromptsDir(cand)) return cand;
            Path parent = cur.getParent();
            if (parent == null || parent.equals(cur)) break;
            cur = parent;
        }
        return null;
    }

    public static Path resolveDir() {
        if (dirValue != null) return dirValue;
        synchronized (DIR_LOCK) {
            if (dirValue != null) return dirValue;
            List<String> attempted = new ArrayList<>();
            String env = System.getenv("CODEX_WEB_PROMPTS");
            if (env != null && !env.isEmpty()) {
                attempted.add(env);
                Path ep = Paths.get(env);
                if (isPromptsDir(ep)) {
                    dirValue = ep;
                    return dirValue;
                }
            }
            // From jar/class location.
            try {
                Path src = Paths.get(Prompts.class.getProtectionDomain()
                    .getCodeSource().getLocation().toURI());
                Path start = Files.isDirectory(src) ? src : src.getParent();
                if (start != null) {
                    Path found = walkUp(start, attempted);
                    if (found != null) {
                        dirValue = found;
                        return dirValue;
                    }
                }
            } catch (URISyntaxException | NullPointerException ignored) {
                // fall through
            }
            // From cwd.
            Path cwd = Paths.get("").toAbsolutePath();
            Path found = walkUp(cwd, attempted);
            if (found != null) {
                dirValue = found;
                return dirValue;
            }
            throw new IllegalStateException("prompts directory not found; tried: "
                + String.join(", ", attempted));
        }
    }

    public static String render(String toolName, Map<String, String> vars) {
        String tpl = TPL_CACHE.computeIfAbsent(toolName, name -> {
            Path dir = resolveDir();
            try {
                return Files.readString(dir.resolve(name + ".md"));
            } catch (IOException e) {
                throw new IllegalStateException("read prompt " + name + ": " + e.getMessage(), e);
            }
        });
        String out = tpl;
        for (Map.Entry<String, String> e : vars.entrySet()) {
            out = out.replace("{" + e.getKey() + "}", e.getValue());
        }
        return out;
    }
}
